doc/source/data/key-concepts.rst

```diff
@@ -57,6 +57,122 @@ Dataset Execution Model
 
 This page overviews the execution model of Datasets, which may be useful for understanding and tuning performance.
 
+By default, all Datasets operations are eager (except for data reading, which is
+semi-lazy; see the :ref:`deferred reading docs <dataset_deferred_reading>`), executing
+each stage synchronously. This provides a simpler iterative development and debugging
+experience, allowing you to inspect up-to-date metadata (schema, row count, etc.) after
+each operation, greatly improving the typical "Getting Started" experience.
+
+However, this eager execution mode can result in less optimal (i.e. slower) execution
+and increased memory utilization compared to what's possible with a lazy execution mode.
+That's why Datasets offers a lazy execution mode, which you can transition to after
+you're done prototyping your Datasets pipeline.
+
+Lazy Execution Mode
```

Comment:
These seem pretty advanced. Why not move them out of key concepts and link to them?

At a minimum, it shouldn't be introduced prior to "Read".

Reply:
Yeah I'm thinking that this will be extracted into a user guide once we de-experimentalize lazy mode. I'll move it below the read section, lmk if you think that's all right for now.

Reply:
Hmm user guide sounds good to me. I took another look and moved it to the end of the doc for now.